overlord/hookstate/ctlcmd/refresh.go

```diff
@@ -122,33 +125,71 @@ type updateDetails struct {
 	restart bool
 }
 
-func getUpdateDetails(context *hookstate.Context) *updateDetails {
+// refreshCandidate is a subset of refreshCandidate defined by snapstate and
+// stored in "refresh-candidates".
+type refreshCandidate struct {
+	Channel     string         `json:"channel,omitempty"`
+	Version     string         `json:"version,omitempty"`
+	SideInfo    *snap.SideInfo `json:"side-info,omitempty"`
+	InstanceKey string         `json:"instance-key,omitempty"`
+}
+
+func getUpdateDetails(context *hookstate.Context) (*updateDetails, error) {
 	context.Lock()
 	defer context.Unlock()
 
 	if context.IsEphemeral() {
 		// TODO: support ephemeral context
-		return nil
+		return nil, nil
 	}
 
 	var base, restart bool
 	context.Get("base", &base)
 	context.Get("restart", &restart)
 
-	// TODO: get revision, version etc. from refresh-candidates.
+	var candidates []*refreshCandidate
+	st := context.State()
+	if err := st.Get("refresh-candidates", &candidates); err != nil {
+		return nil, err
+	}
 
 	up := updateDetails{
 		base:    base,
 		restart: restart,
 	}
-	return &up
+
+	// try to find revision/version/channel info from refresh-candidates; it
+	// may be missing if the hook is called for snap that is just affected by
+	// refresh but not refreshed itself, in such case this data is not
+	// displayed.
+	for _, cand := range candidates {
+		instanceName := snap.InstanceName(cand.SideInfo.RealName, cand.InstanceKey)
+		if instanceName == context.InstanceName() {
+			up.channel = cand.Channel
+			up.revision = cand.SideInfo.Revision
+			up.version = cand.Version
+			break
+		}
+	}
+	// refresh-hint not present, look up channel info in snapstate
+	if up.channel == "" {
+		var snapst snapstate.SnapState
+		if err := snapstate.Get(st, context.InstanceName(), &snapst); err != nil {
+			return nil, fmt.Errorf("internal error: cannot get snap state for %q: %v", context.InstanceName(), err)
+		}
+		up.channel = snapst.TrackingChannel
+	}
+	return &up, nil
 }
 
-func (c *refreshCommand) printPendingInfo() {
-	details := getUpdateDetails(c.context())
+func (c *refreshCommand) printPendingInfo() error {
+	details, err := getUpdateDetails(c.context())
+	if err != nil {
+		return err
+	}
 	// XXX: remove when ephemeral context is supported.
 	if details == nil {
-		return
+		return nil
 	}
 	c.printf("pending: %s\n", details.pending)
```

Comment:
we don't have yet code to set pending is that right?

Reply:
Right. I've added a TODO (I could handle "pending: ready|none" right now but would prefer to address it together with "inhibited") in a followup.